### DISCUSSION TITLE
Seeking Guidance on Proper Execution of calcGrains for Partially Imperfect EBSD Data

### DISCUSSION THREAD

**Discussion by PengWei97** [mtex-toolbox/mtex#2238] [Ask Anything]
Dear METX GitHub Community,

I hope this message finds you well. I am currently working with an EBSD data file that is not entirely perfect. Specifically, the IPF map has areas with insufficient resolution in the upper part, making it unsuitable for using calcGrains. And the lower part of the map has sufficient data and is suitable for grain identification using calcGrains.

Unfortunately, when I execute calcGrains on the entire dataset, the visualization results are unsatisfactory, as shown in the attached image. 
![image](https://github.com/user-attachments/assets/83d1e37b-8ae4-42be-acbe-dd98700b377b)

My goal is to apply calcGrains only to the regions with enough pixel density (the lower part) and avoid applying it to the sparse areas (the upper part). Could you please provide guidance on how to achieve this? Specifically, I would like to know:
1. How to apply `calcGrains` only to the regions with sufficient data points.
2. Any best practices or tips to ensure accurate grain identification in the targeted areas.

Your expertise and advice would be greatly appreciated. Thank you in advance for your help!

Best regards,

Wei

**Comment by zmichels** [comment]
In this case, you could crop the dataset to the lower region before running calcGrains()

using selectInteractive()

https://mtex-toolbox.github.io/EBSD.selectInteractive.html

**Reply by PengWei97** [reply]
Thank you. I have already tried using the mouse to rotate the rectangular region, but it did not meet my expectations. 

Actually, the region of interest is not a well-defined rectangular area; it has an irregular boundary, as shown in the following figure:
![image](https://github.com/user-attachments/assets/ec7914e4-6adf-46a8-8d3a-d59cc2abcafa)

**Comment by nyyssont** [comment]
You could take a look at the section "Adapting the Alpha Parameter" here:
https://mtex-toolbox.github.io/GrainReconstruction.html

Try:
[grains, ebsd.grainId] = calcGrains(ebsd,'alpha',10,'angle',10*degree);

**Reply by PengWei97** [reply]
Thanks, but it still does not work. I executed the following command, and the results is as show below:
```bash
[grains_phase1, ebsd_phase1.grainId] = calcGrains(ebsd_phase1,'alpha',10,'angle',10*degree);
```
![image](https://github.com/user-attachments/assets/289aed88-a7b4-423a-8d62-d7b2b2e26e02)

And then, 
```bash
ebsd_phase1(grains_phase1(grains_phase1.grainSize < 20)) = [];
[grains_phase1, ebsd_phase1.grainId] = calcGrains(ebsd_phase1,'alpha',10,'angle',10*degree);
```
![image](https://github.com/user-attachments/assets/728f0222-3aa2-46d3-991b-9004ab69abd0)

**Comment by PengWei97** [comment]
So, my current idea is to identify and remove invalid grains by calculating the ratio of the area occupied by valid pixels with each grain to the total area of grain. Specifically, if the radio is below a certain threshold, the should be removed.

I am not sure if this idea is reasonable, or whether MTEX supports such as operates. Specifically, I want to obtain the total area of each grain and the total area occupied by pixels with the grains, and then remove the identified grains if the ratio is below a certain threshold.

Thanks
Wei
